[distributed.nn] Change remote module template instantiator to write to tmp folder

xush6528 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#40173 [distributed.nn] Change remote module template instantiator to write to tmp folder**

- Avoid path sharing across runs and workers, so even the test methods/workers run in parallel on the same host, they don't interfere with each other.
- On some environment (e.g. fb internal CI platform), the torch package file tree is not writable. But the temporary folder chosen by Python `tempfile` module is always writable, on linux it's "/tmp".

close https://github.com/pytorch/pytorch/issues/40120

Differential Revision: [D5708493](https://our.internmc.facebook.com/intern/diff/D5708493/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D5708493/)!

xush6528 committed:
[distributed.nn] Change remote module template instantiator to write to tmp folder

- Avoid path sharing across runs.

Differential Revision: [D5708493](https://our.internmc.facebook.com/intern/diff/D5708493/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D5708493/)!

[ghstack-poisoned]

mrshenli commented:
Will this fix #40120?

mrshenli commented on torch/distributed/nn/jit/instantiator.py:
```diff
@@ -3,13 +3,11 @@
-import pathlib
+import tempfile
 import traceback
+import sys
```
This should be before `tempfile`?

xush6528 replied:
@mrshenli Updated.

xush6528 commented:
> Will this fix #40120?

Likely. I think that error happens if another test run is writing to the same file path, when opening the file, it cleans up the file (python `open(mode="w", ..)`).

xush6528 committed:
Update on "[distributed.nn] Change remote module template instantiator to write to tmp folder"


- Avoid path sharing across runs, so even the test methods run in parallel, they don't interfere with each other.
- On some environment (e.g. fb internal CI platform), the torch package file tree is not writable. But the temporary folder chosen by Python `tempfile` module is always writable, on linux it's "/tmp".

Differential Revision: [D5708493](https://our.internmc.facebook.com/intern/diff/D5708493/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D5708493/)!

[ghstack-poisoned]

xush6528 committed:
Update on "[distributed.nn] Change remote module template instantiator to write to tmp folder"


- Avoid path sharing across runs, so even the test methods run in parallel, they don't interfere with each other.
- On some environment (e.g. fb internal CI platform), the torch package file tree is not writable. But the temporary folder chosen by Python `tempfile` module is always writable, on linux it's "/tmp".

close https://github.com/pytorch/pytorch/issues/40120

Differential Revision: [D5708493](https://our.internmc.facebook.com/intern/diff/D5708493/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D5708493/)!

[ghstack-poisoned]

mrshenli commented on torch/distributed/nn/jit/instantiator.py:
```diff
@@ -19,6 +15,10 @@
 
 
 _FILE_PREFIX = "_remote_module_"
+_TEMP_DIR = tempfile.TemporaryDirectory()
```
Would I be correct if  assume each process will have its own temp dir to write/read generated code?

xush6528 replied:
@mrshenli Yes, each process / Python Interpreter / Worker will have its own temp dir.

mrshenli commented on torch/distributed/nn/jit/instantiator.py:
```diff
@@ -62,20 +62,6 @@ def get_arg_return_types_from_interface(module_interface):
 
 
-@atexit.register
-def cleanup_generated_modules():
```
where does this function go? Looks like test is still calling this?

xush6528 replied:
This is handled by destructing of `tempfile.TemporaryDirectory`. Just updated the tests.

mrshenli replied:
Discussed with @xush6528 offline. This is now handled by `TemporaryDirectory()`.

And it will live until the script exists (see https://stackoverflow.com/questions/17301091/python-lifetime-of-module-global-variables). This should be sufficient for now. If we want to delete it earlier, we can add some APIs for that later.

xush6528 committed:
Update on "[distributed.nn] Change remote module template instantiator to write to tmp folder"


- Avoid path sharing across runs and workers, so even the test methods/workers run in parallel on the same host, they don't interfere with each other.
- On some environment (e.g. fb internal CI platform), the torch package file tree is not writable. But the temporary folder chosen by Python `tempfile` module is always writable, on linux it's "/tmp".

close https://github.com/pytorch/pytorch/issues/40120

Differential Revision: [D5708493](https://our.internmc.facebook.com/intern/diff/D5708493/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D5708493/)!

[ghstack-poisoned]

mrshenli reviewed (approved):
LGTM! Let's wait for tests before landing.

mrshenli commented:
Hmm, since the global var will live throughout the lifetime of the process, it looks like our current test can no longer verify the folder is deleted properly? I wonder if we should add a test that uses subprocess to run the `instantiator` and let the main process to test the temp file deletion?

xush6528 commented:
> he process, it looks like our current test can no longer verify the folder is deleted properly? I wonder if we should add a test that uses subprocess to run the `instantiator` and let the

I don't think we need to add tests for python library. That's external to pytorch repo.

mari77111 commented:
El mié., 17 jun. 2020 21:59, Shihao Xu <notifications@github.com> escribió:

> *@xush6528* commented on this pull request.
> ------------------------------
>
> In torch/distributed/nn/jit/instantiator.py
> <https://github.com/pytorch/pytorch/pull/40173#discussion_r441798329>:
>
> > @@ -19,6 +15,10 @@
>
>
>  _FILE_PREFIX = "_remote_module_"
> +_TEMP_DIR = tempfile.TemporaryDirectory()
>
> @mrshenli <https://github.com/mrshenli> Yes, each Process / Python
> Interpreter / Worker will have its own temp dir.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pytorch/pytorch/pull/40173#discussion_r441798329>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AN44JL6E6J43N4LKXOFFIX3RXEOARANCNFSM4OA2RSDQ>
> .
>
